src/librustc/ty/layout.rs

```diff
@@ -511,70 +519,160 @@ pub struct Struct {
     /// If true, the size is exact, otherwise it's only a lower bound.
     pub sized: bool,
 
-    /// Offsets for the first byte of each field.
+    /// Offsets for the first byte of each field, ordered to match the source definition order.
+    /// This vector does not go in increasing order.
     /// FIXME(eddyb) use small vector optimization for the common case.
     pub offsets: Vec<Size>,
 
+    /// Maps source order field indices to memory order indices, depending how fields were permuted.
+    /// FIXME (camlorn) also consider small vector  optimization here.
+    pub memory_index: Vec<u32>,
+
     pub min_size: Size,
 }
 
+// Info required to optimize struct layout.
+#[derive(Copy, Clone, Eq, PartialEq, Ord, PartialOrd, Debug)]
+enum StructKind {
+    // A tuple, closure, or univariant which cannot be coerced to unsized.
+    AlwaysSizedUnivariant,
+    // A univariant, the last field of which may be coerced to unsized.
+    MaybeUnsizedUnivariant,
+    // A univariant, but part of an enum.
+    EnumVariant,
+}
+
 impl<'a, 'gcx, 'tcx> Struct {
-    pub fn new(dl: &TargetDataLayout, packed: bool) -> Struct {
-        Struct {
+    // FIXME(camlorn): reprs need a better representation to deal with multiple reprs on one type.
+    fn new(dl: &TargetDataLayout, fields: &Vec<&'a Layout>,
+                  reprs: &[attr::ReprAttr], kind: StructKind,
+                  scapegoat: Ty<'gcx>) -> Result<Struct, LayoutError<'gcx>> {
+        let packed = reprs.contains(&attr::ReprPacked);
+        let mut ret = Struct {
             align: if packed { dl.i8_align } else { dl.aggregate_align },
             packed: packed,
             sized: true,
             offsets: vec![],
+            memory_index: vec![],
             min_size: Size::from_bytes(0),
+        };
+
+        if fields.len() == 0 {return Ok(ret)};
```

Comment:
No, this can die.  I think the discussion in which it was pointed out to me that empty vecs don't allocate happened later.